### ISSUE TITLE
Wrong CA1508 when struct has a userd-defined conversion to nullable reference type

### ISSUE THREAD

**Issue by JeppeSN** [dotnet/roslyn-analyzers#6435]
### Analyzer

**Diagnostic ID**: [CA1508](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1508): `Avoid dead conditional code`

### Analyzer source

**NuGet Package**: [Microsoft.CodeAnalysis.NetAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers)

**Version**: 7.0.102

### Describe the bug

When a `struct` has a user-defined conversion (either `explicit operator` or `implicit operator`) to a _nullable_ reference type, the CA1508 emits false positives. This was noted when using `StackExchange.Redis.RedisValue` which has such an operator, but for simplicity is reproduced with struct `S` below.

### Steps To Reproduce

Make a C# program like this:
```csharp
#nullable enable

public readonly struct S {
    public int I { get; init; }

    public static explicit operator byte[]?(S s) {
        if (s.I < 0) {
            return null;
        }
        return new byte[s.I];
    }
}

public static class C {

    public static bool M(S s) {
        byte[]? a = (byte[]?)s;
        bool b = a == null;      // incorrect CA1508; this b could be true or false
        return b;
    }

    private static void Main() {
        bool r = M(new S { I = -1, });
    }
}
```
In method `M`, the Boolean `b` could clearly be true, but CA1508 claims it is always false.

Also happens with `implicit operator` (like in `RedisValue` struct).

### Expected behavior
No code analysis error.

### Actual behavior
False positive `Error CA1508 'a == null' is always 'false'. Remove or refactor the condition(s) to avoid dead code.` emitted from the line where `b` is declared.
